OneSignalSDK/onesignal/src/main/java/com/onesignal/OSSessionManager.java

```diff
@@ -0,0 +1,295 @@
+                return this;
+            }
+
+            public SessionResult build() {
+                return new SessionResult(this);
+            }
+        }
+    }
+
+    interface SessionListener {
+        // Fire with the SessionResult that just ended.
+        void onSessionEnding(@NonNull SessionResult lastSessionResult);
+    }
+
+    public enum Session {
+        DIRECT,
+        INDIRECT,
+        UNATTRIBUTED,
+        DISABLED,
+        ;
+
+        public boolean isDirect() {
+            return this.equals(DIRECT);
+        }
+
+        public boolean isIndirect() {
+            return this.equals(INDIRECT);
+        }
+
+        public boolean isAttributed() {
+            return this.isDirect() || this.isIndirect();
+        }
+
+        public boolean isUnattributed() {
+            return this.equals(UNATTRIBUTED);
+        }
+
+        public boolean isDisabled() {
+            return this.equals(DISABLED);
+        }
+
+        public static @NonNull Session fromString(String value) {
+            if (value == null || value.isEmpty())
+                return UNATTRIBUTED;
+
+            for (Session type : Session.values()) {
+                if (type.name().equalsIgnoreCase(value))
+                    return type;
+            }
+            return UNATTRIBUTED;
+        }
+    }
+
+    @NonNull protected Session session;
+    @Nullable private String directNotificationId;
+    @Nullable private JSONArray indirectNotificationIds;
+    @NonNull private SessionListener sessionListener;
+
+    public OSSessionManager(@NonNull SessionListener sessionListener) {
+        this.sessionListener = sessionListener;
+        this.initSessionFromCache();
+    }
+
+    private void initSessionFromCache() {
+        session = OutcomesUtils.getCachedSession();
+        if (session.isIndirect())
+            indirectNotificationIds = getLastNotificationsReceivedIds();
+        else if (session.isDirect())
+            directNotificationId = OutcomesUtils.getCachedNotificationOpenId();
+
+    }
+
+    void addSessionNotificationsIds(@NonNull JSONObject jsonObject) {
+        if (session.isUnattributed())
+            return;
+
+        try {
+            if (session.isDirect()) {
+                jsonObject.put(DIRECT_TAG, true);
+                jsonObject.put(OutcomesUtils.NOTIFICATIONS_IDS, new JSONArray().put(directNotificationId));
+            }
+            else if (session.isIndirect()) {
+                jsonObject.put(DIRECT_TAG, false);
+                jsonObject.put(OutcomesUtils.NOTIFICATIONS_IDS, indirectNotificationIds);
+            }
+        } catch (JSONException e) {
+            OneSignal.Log(OneSignal.LOG_LEVEL.ERROR, "Generating addNotificationId:JSON Failed.", e);
+        }
+    }
+
+    void restartSessionIfNeeded() {
+        // Avoid reset session if app was focused due to a notification click (direct session recently set)
+        if (OneSignal.getAppEntryState().isNotificationClick())
```

Comment:
Good point, I think should revisit it later however. We have a ton of code today looking at OneSignal.java to get global state and need to come up with a good pattern for it. OneSignal.java should really be just incharge of holding onto managers / controllers and have other classes for states where they fit best. When we add these other class we will do need to prevent circular references which is what we have here though.

Reply:
Eventually we agree we need to create independent classes from OneSignal that handle these states